runners/spark/src/main/java/org/apache/beam/runners/spark/io/MicrobatchSource.java

```diff
@@ -86,12 +80,23 @@
     this.readerCacheInterval = readerCacheInterval;
   }
 
+  private static synchronized void initReaderCache(final long readerCacheInterval) {
+    if (readerCache == null) {
+      LOG.info("Creating reader cache. Cache interval = " + readerCacheInterval + " ms.");
+      readerCache =
+          CacheBuilder.newBuilder()
+                      .expireAfterAccess(readerCacheInterval, TimeUnit.MILLISECONDS)
+                      .removalListener(new ReaderCacheRemovalListener())
+                      .build();
+    }
+  }
+
   /**
    * Divide the given number of records into {@code numSplits} approximately
    * equal parts that sum to {@code numRecords}.
    */
-  private static long[] splitNumRecords(long numRecords, int numSplits) {
-    long[] splitNumRecords = new long[numSplits];
+  private static long[] splitNumRecords(final long numRecords, final int numSplits) {
+    final long[] splitNumRecords = new long[numSplits];
```

Comment:
these `final` don't do anything.

You typically want `final` for member variables that should not change after construction, and for variables you're going to pass into anonymous inner classes.